File: program1/c/main.c
#include "stdio.h"
#include "string.h"
#include "pthread.h"
#include "stdlib.h"
#include <unistd.h>

struct Process {
    long threadId;
    char processName;
    long waitTime;
};

void *Scheduled_Process(void *process) {
    // Greet
    struct Process *p = (struct Process *)process;
    FILE *outputFile;
    outputFile = fopen("OutputFile", "a");
    fprintf(outputFile, "Thread%ld: %s begins\n", p->threadId, &p->processName);
    fclose(outputFile);

    // Loop wait time
    int numOfLoops = p->waitTime / 5;
    for (int i = 0; i < numOfLoops; i++) {
        outputFile = fopen("OutputFile", "a");
        fprintf(outputFile, "Thread%ld: Loop %d of %d loops total\n", p->threadId, i, numOfLoops);
        fclose(outputFile);
        usleep(5000);
    }

    // Close
    outputFile = fopen("OutputFile", "a");
    fprintf(outputFile, "Thread%ld: %s ends\n", p->threadId, &p->processName);
    fclose(outputFile);
    pthread_exit(NULL);
}

int main() {
    // Get input file
    char fileLocation[128];
    strcpy(fileLocation, "./InputFile");
    FILE *inputFile = fopen(fileLocation,"r");
    if (inputFile == NULL) {
        printf("Unable to open file.\n");
        return -1;
    }

    // Get file contents
    char lineContent[32];
    struct Process processes[32];
    int index = 0;
    while (fgets(lineContent, sizeof(lineContent), inputFile)) {
        struct Process newProcess;
        char *token = strtok(lineContent, " "); // First Token of the line
        strcpy(&newProcess.processName, token);
        token = strtok(NULL, " "); // Second Token of the line
        char *endptr;
        newProcess.waitTime = strtol(token, &endptr, 10);
        processes[index] = newProcess;
        index++;
    }

    // Create the threads and wait the time
    int numOfThreads = index;
    int referenceCounter;
    pthread_t threads[numOfThreads];

    for (long i = 0; i < numOfThreads; i++) {
        processes[i].threadId = i + 1; // Starting at 1
        referenceCounter = pthread_create(&threads[i], NULL, Scheduled_Process, (void *)&processes[i]);
        if (referenceCounter) {
            printf("Error\n");
            exit(-1);
        }
    }
    pthread_exit(NULL);
    return 0;
}

File: program1/rust/src/main.rs
use core::time;
use std::fs::{File, OpenOptions};
use std::io::{self, BufRead, Write};
use std::path::Path;
use std::sync::{Arc, Mutex};
use std::thread::{self, sleep};

/**
 * Program written by Jason Carnahan as a Rust version of the CS 690 program 1.
 * See the `program1.pdf` requirements document for details.
 *
 * The program limits the use of the `.unwrap()` for the sake of custom error handling.
 */
fn main() {
    // Get input file and open for reading
    let file_location = Path::new("InputFile");
    let file = match File::open(&file_location) {
        Ok(file) => file,
        Err(why) => panic!("Couldn't open {}: {}", file_location.display(), why),
    };

    // Read in the lines of the file, parse name and wait time, and store in a vector
    let mut process_vector: Vec<Process> = Vec::new();
    let lines = io::BufReader::new(file).lines();
    for (index, line) in lines.flatten().into_iter().enumerate() {
        let split: Vec<&str> = line.split(" ").collect();

        let process = Process {
            thread_id: index + 1,
            process_name: split[0].to_string(),
            wait_time: match split[1].parse::<i32>() {
                Ok(num) => num,
                Err(_) => panic!("Could not parse wait time."),
            },
        };

        process_vector.push(process);
    }

    // Open the output file and create and mutex that can safely pass among the threads
    let output_file = OpenOptions::new()
        .create(true)
        .append(true)
        .open("OutputFile")
        .expect("Failed accessing output file");
    let writer = Arc::new(Mutex::new(output_file));

    // Spawn a thread for each "process" and print the data using a Mutex
    let num_of_threads = process_vector.len();
    let mut thread_handles = vec![]; // Storage to wait for each thread to finish
    for i in 0..num_of_threads {
        let writer = Arc::clone(&writer); // A mutex for each thread
        let process = process_vector[i].clone();
        let thread_handle = thread::spawn(move || {
            scheduled_process(process, &writer);
        });
        thread_handles.push(thread_handle);
    }

    for thread in thread_handles {
        thread.join().unwrap();
    }
}

#[derive(Clone)]
struct Process {
    thread_id: usize,
    process_name: String,
    wait_time: i32,
}

fn scheduled_process(process: Process, writer: &Mutex<File>) {
    // Greet the start of the thread
    writer
        .lock()
        .unwrap()
        .write(
            format!(
                "Thread{}: {} begins\n",
                process.thread_id, process.process_name
            )
            .as_bytes(),
        )
        .expect("Could not write to outputfile.");

    // Loop for the wait time and print status
    let num_of_loops = process.wait_time / 5;
    for i in 1..num_of_loops + 1 {
        writer
            .lock()
            .unwrap()
            .write(
                format!(
                    "Thread{}: Loop {} of {} loops total\n",
                    process.thread_id, i, num_of_loops
                )
                .as_bytes(),
            )
            .expect("Could not write to outputfile.");
        sleep(time::Duration::from_micros(5000));
    }

    // End the thread
    writer
        .lock()
        .unwrap()
        .write(
            format!(
                "Thread{}: {} ends\n",
                process.thread_id, process.process_name
            )
            .as_bytes(),
        )
        .expect("Could not write to outputfile.");
}
